docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,312 @@
+Let's consider a bear to be dependent on a project bear ``Fizz`` and a file bear
+``Buzz`` then the corresponding DependencyBear let's call it ``FizzBuzz`` will
+look like the following:
+
+::
+
+    class FizzBuzzBear(DependencyBear):
+        BEAR_DEPS = {FizzBear, BuzzBear}
+
+        def analyze(self, dependency_bear, dependency_result, a_number=100):
+            yield '{} ({}) - {}'.format(
+                dependency_bear.name, a_number, dependency_result)
+
+Ability to modify bear dependencies at runtime
+----------------------------------------------
+
+A bear might depend on multiple bears (to run) before its execution can begin.
+``Bear.BEAR_DEPS`` is just a collection (a set) of classes (Bear classes) that
+need to be executed before that bear can run. Once all these dependencies
+(Bears) have run their results are passed into ``self.dependency_results``.
+The results are in the form of a dictionary with the types of the bears and
+their corresponding results (in the form of a list) as *Key-Value* pairs.
+From the previous example if we try to access the BEAR_DEPS of the ``BarBear``
+we will get the result ``{<class 'coalib.core.Bear.FooBear'>}``.
+
+In the `__init__()` method of the class ``Bear.py`` the dependencies specified
+in the ``BEAR_DEPS`` are copied to every instance of a Bear run using
+``self.BEAR_DEPS = set(self.BEAR_DEPS)`` which makes runtime modifications
+possible.
+
+Override bears
+--------------
+
+A NextGen bear has to have the following functions to perform analysis:
+
+- ``analyze``: This method contains the code that performs the actual code
+  analysis routine that that bear is used for.
+
+- ``generate_tasks``: This method is a part of the parent ``Bear.py`` class
+  and returns tuples containing the positional arguments as a tuple and the
+  keyword arguments in the form of a dict. These are actually the task objects
+  that are scheduled and executed by the core. An absence of this method raises
+  ``NotImplementedError``. For example if we instantiate an object of
+  ``FileBear`` and call the ``generate_tasks`` method we would get a tuple of a
+  tuple of the filename and the file in the file_dict and a dictionary of kwargs
+  to the ``analyze`` method.
+
+  ::
+
+      from coalib.core.FileBear import FileBear
+
+      section = Section('test-section')
+      file_dict = {'fileA.py': []}
+
+      FileBear(section, file_dict).generate_tasks()
+
+In the same way ``generate_tasks`` can be used in bears inheriting from
+``ProjectBear`` or ``DependencyBear``.
+
+Let's write our own bears with custom ``generate_tasks`` methods.
+
+::
+
+    # This bear provides some code analysis
+    class SomeBear(Bear):
+
+        def analyze(self, bear, section_name, file_dict):
+            yield "Some analysis"
+
+    # This bear depends on the above bear and performs some
+    # more analysis after receving its results
+    class SomeOtherBear(Bear):
+        BEAR_DEPS = {SomeBear}
+
+        def analyze(self, *args):
```

Comment:
Don't do `*args`, if possible write them out so people can easier see the commonality between `analyze` and `generate_tasks`